### ISSUE TITLE
Interesting work, future continuation?

### ISSUE THREAD

**Issue by chri002** [joaopauloschuler/less-parameters-llm#1]
Really interesting and probably innovative work, one can say a real optimization for LLM.
I wanted to know if you were thinking of also bringing "optimizations" to slightly larger models like the 4B, 7B or who knows even 14/20B, and if in that case the training does not go to degrade.
Otherwise great, this could really open the way for LLMs on low memory devices, without excessive quantizations.

In addition, you could add some of the usual benchmarks between the three "versions" to see if the model eventually degrades or even improves.

**Comment by joaopauloschuler** [comment]
Dear @chri002 ,
You have asked intelligent questions that deserve a reply.

regarding
> I wanted to know if you were thinking of also bringing "optimizations" to slightly larger models like the 4B, 7B or who knows even 14/20B, and if in that case the training does not go to degrade.

If you like, you can try yourself. This is how:
1) Clone [the experiment](https://github.com/joaopauloschuler/less-parameters-llm/blob/main/raw/JP47D56_kphi3_3T.ipynb) into your Google Colab (or other) environment, and then change this portion of code:
```
EMBED_SIZE = 3072
HIDDEN_SIZE = 3072
HEADS = HIDDEN_SIZE // 96
torch.set_default_dtype(torch.float32)
model = False
training_args = False
trainer = False
generator = False
torch.cuda.empty_cache()
# """
custom_config = KPhi3Config(
    vocab_size=VOCAB_SIZE,  # Adjust as needed,
    pad_token_id=32000,
    bos_token_id=1,
    eos_token_id=32000,
    embed_size=EMBED_SIZE,
    hidden_size=HIDDEN_SIZE, # embedding size
    num_hidden_layers=3,  # Number of decoder layers
    num_attention_heads=HEADS,
    ffn_dim=HIDDEN_SIZE,  # Dimension of the feed-forward layer,
    intermediate_size=HIDDEN_SIZE*3,  # Size of the intermediate feed-forward layer
    max_position_embeddings=131072,
    # do_layer_norm_before=True,
    min_channels_per_group=256,
    attn_implementation="flash_attention_2",
    # sliding_window=256,
    torch_dtype=torch.float32
    # word_embed_proj_dim=512,
    # Add any other parameters you want to customize
)
model = KPhi3ForCausalLM(config=custom_config)
```
I would suggest to change just the "num_hidden_layers" up to 32 layers as the original phi-3 has 32 layers. Also, change the `REPO_NAME` constant if you are planning to push the result to hf. Then, you can train by yourself and please share the result :) .

From my own experience, for the past 6 years with this kind of optimization ( [see here](https://www.researchgate.net/profile/Joao-Paulo-Schwarz-Schuler) ), generally, I can say that the optimized versions are more stable at training than their dense counterparts.

I suspect that the 3 experiments (neural networks) described in the paper are too dumb for a modern benchmark. Therefore, I'm just looking at the loss.

This technical report belongs to a work in progress. There is plenty more to come :) .

Kind regards,
JP.

**Comment by joaopauloschuler** [comment]
@chri002 , if you share the amount of RAM, the config and the error message, I can try to help.
